### PR TITLE
Ref farming

### PR DESCRIPTION
**Pull request by marco-sundsk**
# Ref Farming Contract Merge PR
This contract has been implemented and tested with ref-finance contract (as a mft token contract) and test-token contract (as reward token contract). The test code is in tests dir.   
And for reviewers, it is strongly recommended to read `how_does_it_work.md` first.  
To get a clear look of the interface of this contract, please read `readme.md`.   
If you have any question, please don't hesitate to contact me.  
My email: sun.dsk1@gmail.com  
My telegram; https://t.me/MarcoSun80

### CODE REVIEW

**Review by referencedev — COMMENTED** [review]
Started looking at the code, but there are a lot of english spelling mistakes in function names and comments.
If you are using Clion or VS Studio - they have modules with spell checker that can help correct english spelling.

File: ref-farming/tests/test_one_farm_two_farmer.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    call!(farmer1, farming.storage_deposit(None, None), deposit = to_yocto("1"))
+    .assert_success();
+    let out_come = call!(
+        farmer1,
+        pool.mft_transfer_call("0".to_string(), to_va(farming_id()), to_yocto("1").into(), None, "".to_string()),
+        deposit = 1
+    );
+    out_come.assert_success();
+    // println!("{:#?}", out_come.promise_results());
+    show_farminfo(&farming, farm_id.clone());
```
Comment (referencedev) [inline ref-farming/tests/test_one_farm_two_farmer.rs:55]: let's add asserts here for values to make sure these tests can run without human oversight

File: ref-farming/tests/common/mod.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
+use std::convert::TryFrom;
```
Comment (referencedev) [inline ref-farming/tests/common/mod.rs:1]: this doesn't need to be a separate folder if it's just a single file. can be common.rs

File: ref-farming/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+mod storage_impl;
+
+mod actions_of_farm;
+mod actions_of_seed;
+mod actions_of_reward;
+mod view;
+
+/// sodu module is used to debug and testing,
+/// remove this module in release version
+mod sudo;
```
Comment (referencedev) [inline ref-farming/src/lib.rs:36]: let's add `#[cfg(build = "debug")]` though I'm not sure why is this needed for debugging either. If you have full access keys to the contract can in theory deploy new contract that will change owner if needed.

File: ref-farming/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+near_sdk::setup_alloc!();
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+
+    // owner of this contract
+    owner_id: AccountId,
```
Comment (referencedev) [inline ref-farming/src/lib.rs:45]: does this contract need an owner? it doesn't seem like it's used anywhere

File: ref-farming/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(owner_id: ValidAccountId) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            owner_id: owner_id.into(),
+            seeds: UnorderedMap::new(b"s".to_vec()),
```
Comment (referencedev) [inline ref-farming/src/lib.rs:66]: let's use StorageKeys model: https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/lib.rs#L33

File: ref-farming/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            farmer_count: 0,
+            farm_count: 0,
+            reward_info: UnorderedMap::new(b"r".to_vec()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    // use std::convert::TryFrom;
```
Comment (referencedev) [inline ref-farming/src/lib.rs:77]: please remove commented out code.

**Review by referencedev — COMMENTED** [review]
Also please run `cargo fmt`. Clion has helpful feature to run it on save.

File: ref-farming/how_does_it_work.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
+# ref-farming
+
+## Terminology
```
[the remark this answers is not in the document]
Comment (marco-sundsk) [reply ref-farming/how_does_it_work.md:3]: Yeah, I will go over the spell mistake ASAP. And about the docs, for now, it is build to help reviewers to understand how it works in my mind and have an anchor to judge the code. And U are right, this part should be refactor and move into docs repo after review, I think.

File: ref-farming/build.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -e
+
+RUSTFLAGS='-C link-arg=-s' cargo build --target wasm32-unknown-unknown --release
+cd ..
+cp target/wasm32-unknown-unknown/release/ref_farming.wasm ./res/
```
[the remark this answers is not in the document]
Comment (marco-sundsk) [reply ref-farming/build.sh:6]: Roger that. Will do ASAP.

File: ref-farming/tests/common/mod.rs (hunk 2 above)
Comment (marco-sundsk) [reply ref-farming/tests/common/mod.rs:1]: The reason to create a mod folder is that the testing would go over all files in tests fold, which cause an unnecessary testing output for this common file. While putting it in subfolder would avoid this.

File: ref-farming/src/lib.rs (hunk 4 above)
Comment (marco-sundsk) [reply ref-farming/src/lib.rs:43]: The owner is used for the next move. Cause you can see the farm's life cycle is incomplete, we can only create a farm and it will last for ever, like the simple-pool in ref-finance. So, the next version of this farming contract, from my view, is to address that. When all the reward of a farm is distributed and claimed out, then it could be removed from contract states. The way to do that, is to have kind of owner of this contract, act as a backend operator, have the rights to clean outdated farms. And like ref-finance, I am thinking adding upgrade function to the contract, which also need the exist of owner, I think.

File: ref-farming/src/lib.rs (hunk 5 above)
Comment (marco-sundsk) [reply ref-farming/src/lib.rs:64]: Wow, cool, neat code. I will follow that.

File: ref-farming/src/lib.rs (hunk 6 above)
Comment (marco-sundsk) [reply ref-farming/src/lib.rs]: OK, I will clean up the code.

File: ref-farming/src/lib.rs (hunk 3 above)
Comment (marco-sundsk) [reply ref-farming/src/lib.rs]: I have removed this sudo mod, change it to owner.rs. What is this owner mod used for, has explained in next comment.

File: ref-farming/src/lib.rs (hunk 5 above)
Comment (marco-sundsk) [reply ref-farming/src/lib.rs]: I leave the code in comment there, cause the main project using near-rs-sdk 3.0.0, where StorageKeys is not supported. I want this PR is limited in ref-farming sub folder, so I won't change Cargo.html of main project. Once the main project upgrade to 3.1.0, I will refresh the code correspondingly.

File: ref-farming/tests/test_one_farm_two_farmer.rs (hunk 1 above)
Comment (marco-sundsk) [reply ref-farming/tests/test_one_farm_two_farmer.rs]: Fixed.
